File: specs/python/zconcat.py
import vgplot as vg

land = vg.spatial("data/countries-110m.json", layer="land")

blueLongitude = vg.param(-20)
blueLatitude = vg.param(-5)
blueRotate = vg.param([blueLongitude, blueLatitude])
redLongitude = vg.param(-134)
redLatitude = vg.param(25)
redRotate = vg.param([redLongitude, redLatitude])

view = vg.vconcat(
    vg.hconcat(
        vg.slider(
            label="Blue longitude", bind=blueLongitude, min=-180, max=180, step=1
        ),
        vg.slider(label="Blue latitude", bind=blueLatitude, min=-90, max=90, step=1),
    ),
    vg.hconcat(
        vg.slider(label="Red longitude", bind=redLongitude, min=-180, max=180, step=1),
        vg.slider(label="Red latitude", bind=redLatitude, min=-90, max=90, step=1),
    ),
    vg.zconcat(
        vg.plot(
            vg.geo(
                land, geometry=vg.geojson("geom"), fill="steelblue", fill_opacity=0.4
            ),
            vg.sphere(stroke="steelblue"),
            vg.width(420),
            vg.height(420),
            vg.margin(10),
            vg.style("overflow: visible;"),
            vg.projection_type("orthographic"),
            vg.projection_rotate(blueRotate),
        ),
        vg.plot(
            vg.geo(land, geometry=vg.geojson("geom"), fill="tomato", fill_opacity=0.4),
            vg.sphere(stroke="tomato"),
            vg.width(420),
            vg.height(420),
            vg.margin(10),
            vg.style("overflow: visible;"),
            vg.projection_type("orthographic"),
            vg.projection_rotate(redRotate),
        ),
    ),
)

File: specs/ts/zconcat.ts
import { Spec } from '@uwdata/mosaic-spec';

export const spec : Spec = {
  "meta": {
    "title": "Layered Globes",
    "description": "Two globes layered on top of one another with `zconcat`, each rotated by its own pair of sliders, to compare the size and shape of distant landmasses (blue starts on Africa, red on Australia). The sliders sit outside the `zconcat`, so they stay usable while the globes overlap. Requires the DuckDB `spatial` extension.\n"
  },
  "data": {
    "land": {
      "type": "spatial",
      "file": "data/countries-110m.json",
      "layer": "land"
    }
  },
  "params": {
    "blueLongitude": -20,
    "blueLatitude": -5,
    "blueRotate": [
      "$blueLongitude",
      "$blueLatitude"
    ],
    "redLongitude": -134,
    "redLatitude": 25,
    "redRotate": [
      "$redLongitude",
      "$redLatitude"
    ]
  },
  "vconcat": [
    {
      "hconcat": [
        {
          "input": "slider",
          "label": "Blue longitude",
          "as": "$blueLongitude",
          "min": -180,
          "max": 180,
          "step": 1
        },
        {
          "input": "slider",
          "label": "Blue latitude",
          "as": "$blueLatitude",
          "min": -90,
          "max": 90,
          "step": 1
        }
      ]
    },
    {
      "hconcat": [
        {
          "input": "slider",
          "label": "Red longitude",
          "as": "$redLongitude",
          "min": -180,
          "max": 180,
          "step": 1
        },
        {
          "input": "slider",
          "label": "Red latitude",
          "as": "$redLatitude",
          "min": -90,
          "max": 90,
          "step": 1
        }
      ]
    },
    {
      "zconcat": [
        {
          "plot": [
            {
              "mark": "geo",
              "data": {
                "from": "land"
              },
              "geometry": {
                "geojson": "geom"
              },
              "fill": "steelblue",
              "fillOpacity": 0.4
            },
            {
              "mark": "sphere",
              "stroke": "steelblue"
            }
          ],
          "width": 420,
          "height": 420,
          "margin": 10,
          "style": "overflow: visible;",
          "projectionType": "orthographic",
          "projectionRotate": "$blueRotate"
        },
        {
          "plot": [
            {
              "mark": "geo",
              "data": {
                "from": "land"
              },
              "geometry": {
                "geojson": "geom"
              },
              "fill": "tomato",
              "fillOpacity": 0.4
            },
            {
              "mark": "sphere",
              "stroke": "tomato"
            }
          ],
          "width": 420,
          "height": 420,
          "margin": 10,
          "style": "overflow: visible;",
          "projectionType": "orthographic",
          "projectionRotate": "$redRotate"
        }
      ]
    }
  ]
};
